File: examples/python/03_pubsub.py
"""
03_pubsub.py — Publish to a GossipSub topic and subscribe to it.

Two threads: one publishes every second; the other subscribes and prints.

Run:
    python 03_pubsub.py
"""
import os
import threading
import time
from moltmesh import A2AClient

ADDR  = os.getenv("A2A_GRPC_ADDR", "")
TOPIC = "example/hello"

def publisher(client: A2AClient):
    for i in range(5):
        time.sleep(1)
        client.publish(TOPIC, f"ping #{i}")
        print(f"[pub] sent ping #{i}")

def subscriber(client: A2AClient):
    print(f"[sub] listening on topic '{TOPIC}' ...")
    count = 0
    for msg in client.subscribe_topic(TOPIC):
        print(f"[sub] received: {msg.payload.decode()}")
        count += 1
        if count >= 5:
            break

def main():
    pub = A2AClient(ADDR).connect()
    sub = A2AClient(ADDR).connect()

    t_sub = threading.Thread(target=subscriber, args=(sub,), daemon=True)
    t_sub.start()

    publisher(pub)
    t_sub.join(timeout=10)

    pub.close()
    sub.close()

if __name__ == "__main__":
    main()

File: examples/typescript/03_pubsub.ts
/**
 * 03_pubsub.ts — Publish to a GossipSub topic and subscribe to it.
 *
 * Run:
 *   npx tsx 03_pubsub.ts
 */
import { A2AClient } from "../../sdk/typescript/openclaw-plugin/src/client.js";

const ADDR  = process.env["A2A_GRPC_ADDR"] ?? "";
const TOPIC = "example/hello";

async function subscribe(client: A2AClient, count: number) {
  console.log(`[sub] listening on topic '${TOPIC}' ...`);
  let received = 0;
  for await (const msg of client.subscribeTopic(TOPIC)) {
    console.log(`[sub] received: ${Buffer.from(msg.payload).toString("utf8")}`);
    if (++received >= count) break;
  }
}

async function publish(client: A2AClient, count: number) {
  for (let i = 0; i < count; i++) {
    await new Promise(r => setTimeout(r, 1000));
    await client.publish(TOPIC, `ping #${i}`);
    console.log(`[pub] sent ping #${i}`);
  }
}

async function main() {
  const pub = new A2AClient(ADDR);
  const sub = new A2AClient(ADDR);

  const subPromise = subscribe(sub, 5);
  await publish(pub, 5);
  await subPromise;

  pub.close();
  sub.close();
}

main().catch(console.error);
